### PR TITLE
ISLANDORA-1947: Bump Internet Archive Book Reader support to V2.0.2

### PR DESCRIPTION
**Pull request by DiegoPino**
**JIRA Ticket**: (https://jira.duraspace.org/browse/ISLANDORA-1947)

* https://github.com/DiegoPino/islandora_internet_archive_bookreader/pull/3

# What does this Pull Request do?

Switch islandora_internet_archive_bookreader to use the 2.0.2 version of the bookreader library. This provides an updated theme, a number of bug fixes, and usability improvements. It also eliminates the need to maintain our own customized fork of the bookreader library.

# What's new?

* Replaces dependency on [our fork of the Internet Archive BookReader](https://github.com/Islandora/internet_archive_bookreader.git) with [the 2.0.2 version from internetarchive.org](https://github.com/internetarchive/bookreader/tree/2.0.2). *(It has not been tested with more recent 2.x versions, or 3.x)*
* Removes bundled copies of jquery (1.5) and jquery_ui (1.8.10). They are not needed.
* Extensively modifies islandora_book_reader.js to work with the 2.x version of bookreader, and provide Islandora-specific customization that had been part of our fork of the bookreader library.

# How should this be tested?

Because this is a large change to an existing feature, testing should be pretty broad.

* Install/upgrade per the README file. (Clear all caches if upgrading.)
* Configure
  - Consult the README section on configuring this module. Do all the controls work as described, save their settings, etc?
* Navigate to a multiple page book object (create one if necessary) and test display and functionality:
  - Top of the viewer controls:
    - Text popup (shows the text of the currently displayed pages in a popup, if available).
    - Info popup (shows metadata in a popup).
    - Sharing popup (shows social media and mail sharing links).
    - Zoom in and out.
    - Search inside the book. Pick a word that you know occurs several times within the book. 
    - Full screen display.
  - Bottom of the viewer controls:
    - One-up, two-up and tiled display. Do the pages display correctly, showing edge to edge on all four sides?
    - Forward and back buttons.
    - Show/hide controls button.
- Testing is needed for both djatoka and IIIF implementations. The main trick for IIIF is making sure you have the IIIF Identifier set correctly. 

# Additional Notes:

* Does this change require documentation to be updated?
  *Yes. This PR already includes a minor change to the README regarding how to install the bookreader library.*
* Does this change add any new dependencies?
  *No.*
* Does this change require any other modifications to be made to the repository (ie. Regeneration activity, etc.)?
  *No.*
* Could this change impact execution of existing code?
  *No. This only affects the display of book objects.*

# PS: 
* We removed some additional features that are in Diego's 7.x-2dev branch which are not strictly necessary to support the newer bookreader library (We propose that these features be evaluated separately in the future):
  - Default image width
  - Info tab view
  - Mobile header

Credits go to: @giancarlobi @patdunlavey @DiegoPino (myself)
Final note: we will have to keep the old fork around for people running older versions. Maybe add a note there too?

# Interested parties
@patdunlavey @giancarlobi @DigitLib  @jonathangreen

### CODE REVIEW

**Comment by DiegoPino** [comment]
@giancarlobi @patdunlavey @digitlib this is in. I made sure we preserved the whole commit history, even if a little bit messy. (that is our reality, let's embrace it). @Islandora/7-x-1-x-committers any one wants to give this a review/try?

**Comment by manez** [comment]
Just a note that everyone involved is covered by a CLA or cCLA, so it's all good on that front!

**Comment by giancarlobi** [comment]
@DiegoPino many thanks to make this happens!

**Comment by DigitLib** [comment]
@DiegoPino GREAT AND THANK YOU!!!!!

**Comment by DigitLib** [comment]
@DiegoPino @patdunlavey @giancarlobi It work with 2.1 also, it has minor changes:
    Add auto mode to 1up autosize (in addition to height and width)
    Remove the old responsiveAutofit (which is not needed anymore)

**Comment by DigitLib** [comment]
@DiegoPino @patdunlavey When I try to open /admin/islandora/search/islandora_solr it only past text  "islandora_solr&page" to bookreader searchbox but not open overlay to configure solr.

Text popup show text from right page below the text from left page, it can be solved with lowering width in css/islandora_internet_archive_bookreader.css (.textRight and .textLeft from 50% to 46%), I can PR this?

**Comment by DiegoPino** [comment]
Andrija. Not sure what you mean with when:
i try to open /admin/islandora/search/islandora_solr it only past text "islandora_solr&page" to bookreader searchbox but not open overlay to configure solr. Is this an error? Does Search work still? Or is this related to something else?
About the CSS, perfect, i can tune that.

**Comment by DigitLib** [comment]
@DiegoPino Searching work fine but when I try to open solr config http://localhost:8000/admin/islandora/search/islandora_solr overlay don't show, but in searchbox of bookreader got the text **islandora_solr&page**. It happens when I'm on object page with bookreader.

File: js/islandora_book_reader.js
Comment (DiegoPino) [inline js/islandora_book_reader.js]: @patdunlavey @giancarlobi @DigitLib this seems off. We had cover support enabled. @patdunlavey do you remember why this was removed during the rebase?

**Comment by DiegoPino** [comment]
@DigitLib any messages in your Browser's JS console? Some JS is playing tricks with Drupal's overlay. For V4 we had to add https://github.com/giancarlobi/islandora_internet_archive_bookreader/blob/7.x-v4Dev/js/jquery_dollar.js but probably not the case here.

**Comment by DigitLib** [comment]
@DiegoPino console don't show anything...

**Comment by DiegoPino** [comment]
@DigitLib ok, i will give this a try tomorrow. I need to re-introduce some code that was removed during the last commit from Pat too. Will let you know

**Comment by DigitLib** [comment]
Great!!!!

**Comment by DigitLib** [comment]
@DiegoPino I catch the problem with solr admin overlay (http://localhost:8000/islandora/object/islandora%3A1#overlay=admin/islandora&page/1/mode/2up=)!
Bookreader capture search word phrase from  http://localhost:8000/admin/islandora**/search/islandora_solr** and add it to searchbox. I changed the links in   https://github.com/Islandora/islandora_solr_search/blob/e5cd8aa93ff1f7f0d4903f28cce5daaeaef61054/islandora_solr.module#L42 and it work so this is the problem...
Here is imgs
solr1
solr2

**Comment by DiegoPino** [comment]
@DigitLib i wonder if this is related to 
https://github.com/internetarchive/bookreader/commit/13ab3ed19aa8579d66e041d469f442518115401a
But also to the fact that previously we have our own 
IslandoraBookReader.prototype.updateLocationHash = function()

I just made a commit to "try" to bring the old function back. I so sadly have no time to test deeply but i would suggest you try to debug this: https://github.com/internetarchive/bookreader/commit/13ab3ed19aa8579d66e041d469f442518115401a and also/or test if recovering that function i just committed helps with the issue (i doubt it to be honest)

**Comment by DiegoPino** [comment]
Something is wrong with Github. I just made a commit and nothing here. Got 2x 500 errors...

**Comment by DiegoPino** [comment]
There it is...

**Comment by MarcusBarnes** [comment]
@DiegoPino From https://www.githubstatus.com/:  "Investigating - We are investigating elevated error rates for Web and API requests.  Jul 18, 18:00 UTC"

**Comment by DigitLib** [comment]
@DiegoPino Tested but this remove many functions, details don't show, open islandora admin page without overlay and most important don't pass searched word from simple search in bookreader to see its location on page... I look for a solution to ignore somehow path http://localhost:8000/admin/islandora/search/islandora_solr

**Comment by DiegoPino** [comment]
My fault @DigitLib there is an issue with the JS. Give me a few hours to correct.

**Comment by DigitLib** [comment]
@DiegoPino OK

**Comment by whikloj** [comment]
Good catch @DigitLib. I did not catch that in my run through. Doing some more testing now.

**Comment by DigitLib** [comment]
@whikloj Great! Good luck!

**Comment by DiegoPino** [comment]
@DigitLib @whikloj @patdunlavey issue is here
https://github.com/internetarchive/bookreader/blob/2.0.2/BookReader/BookReader.js#L4963
Seems like the way the overlay interacts with the location hash arguments + the fact that islandora solr search contains the word "search" makes Bookreader think its a search agument passed via URL

Options are:

Override `BookReader.prototype.paramsFromFragment = function(urlFragment) {`

OR

Remove/override the location polling that overwatches the changes on the location hash

https://github.com/internetarchive/bookreader/blob/2.0.2/BookReader/BookReader.js#L5141-L5169

If people want this in, i could need some help with deciding, quite busy lately and trying to find time to fix this thing. Seems like we are running out of time.

D

File: islandora_internet_archive_bookreader.drush.inc
Comment (DiegoPino) [inline islandora_internet_archive_bookreader.drush.inc]: This is wrong. original dir should be 'bookreader-2.0.2' @patdunlavey do you have time to make a pull here?

**Comment by whikloj** [comment]
@DiegoPino both of those options seem to mean we need to fork IABV again, am I correct or can this be done another way?

**Comment by DiegoPino** [comment]
@whikloj i "think" there is no need to fork anything. If there is, i totally quit here...

I could , e.g override  this (where the whole mess is happening)

```JS
// startLocationPolling
//________
// Starts polling of window.location to see hash fragment changes
BookReader.prototype.startLocationPolling = function() {
    var self = this; // remember who I am
    self.oldLocationHash = window.location.hash;

    if (this.locationPollId) {
        clearInterval(this.locationPollID);
        this.locationPollId = null;
    }

    this.locationPollId = setInterval(function() {
        var newHash = window.location.hash;
        if (newHash != self.oldLocationHash && newHash != self.oldUserHash) {
            // Only process new user hash once
            //console.log('url change detected ' + self.oldLocationHash + " -> " + newHash);
            self.ttsStop();

            // Queue change if animating
            if (self.animating) {
                self.autoStop();
                self.animationFinishedCallback = function() {
                    self.updateFromParams(self.paramsFromFragment(newHash));
                }
            } else { // update immediately
                self.updateFromParams(self.paramsFromFragment(newHash));
            }
            self.oldUserHash = newHash;
        }
    }, 500);
};
```

With a
```JS
IslandoraBookReader.prototype.startLocationPolling()(
````

where i check if the new newHash contains `#overlay` and if so, simply skip the polling.

But as you say, if that does not work, then well, it was good try. Too many changes in a single old branch and IAbookreader 1.x-3.x is quite messy (you can quote me on that).  When did you say code freeze was happening? 😭

**Comment by whikloj** [comment]
@DiegoPino code freeze is today and I need to build the release VM sometime on Sunday as I'm on vacation (starting tomorrow) and heading out of town Monday-Wednesday. I don't want to delay the testers/documenters if I can avoid it.

If you can give the override a try that would be great. I'll see if I can give what you are suggesting an attempt as well.

**Comment by DiegoPino** [comment]
@whikloj @DigitLib i could also check the scope of that interval. If it lives in a property on the main Object (or i can fetch it somehow) i could do a clearInterval(blablathingiwanttodiedie). Was that polling even a feature of the old one?
Also, i just added a comment to the drush.inc . Seems like some of the code committed over my original has flaws (like wrong name for the zip?) so, yeah, not sure how much i can do, i have my hands full with 4 other projects today. Vacations? That sounds fancy and nice (so jealous...)

**Comment by DiegoPino** [comment]
@whikloj but wait. This whole stuff is a bug fix right? (see jira ticket), one that can only be fixed by updating to 2.0.2 (the resize) thing. I know its a semantic distinction, but  looking at how things are, even if i fix this we are out of 24 hours of waiting before merge and i know some folks will have opinions just, just before merging.....
Also, even if we/i had more time, i'm overcommitted for the next few months. I will give this a last try after lunch (1 hour) and lets see how that goes

**Comment by DigitLib** [comment]
@DiegoPino @whikloj @patdunlavey  maybe !? to leave solr thing as a bug? Everything else works excellent!!! And to solve it for next release???

**Comment by whikloj** [comment]
Ok I add this to the bottom of `islandora_book_reader.js` and it seemed to work. 
```
  IslandoraBookReader.prototype.paramsFromFragment = function(urlFragment) {

    if (/#overlay=/.test(urlFragment)) {
      // We are on the overlay so exit.
      return;
    }

    return BookReader.prototype.paramsFromFragment(urlFragment)
  };
```

But if I submit a PR then I can't merge. So if someone wants to test my change and if they think it is good submit a PR to the PR that would be good.

**Comment by DiegoPino** [comment]
@whikloj great! Was writing almost the same code and had the js console ready to test. Give me 15 minutes and i will have that done.

**Comment by DigitLib** [comment]
@whikloj @DiegoPino GREAT!! I'll 'be on comp for 2 hours and try it also...

**Comment by DiegoPino** [comment]
@whikloj++ ok. Added the code, fixed the drush, pulled, tested, checked, no errors in the console and no fragments in the search box.

File: islandora_internet_archive_bookreader.drush.inc
Comment (DiegoPino) [reply islandora_internet_archive_bookreader.drush.inc]: Fixed via https://github.com/Islandora/islandora_internet_archive_bookreader/pull/133/commits/49eb72d1feda9f64731acb3218a5c59146461a21

**Comment by whikloj** [comment]
That works for me 👍 

My only question is related to the "Share this book" is it expected that the "Embed a mini book reader" says "Embed code not currently supported."?

**Comment by whikloj** [comment]
Actually the "Finished" button is not working for me on that pop-up either. The top-right X works, but not "Finished"

**Comment by whikloj** [comment]
Looks like in the one that works they use 
```
onclick="parent.jQuery.colorbox.close();"
```
and on the one that doesn't they use
```
onclick="$.fn.colorbox.close();"
```

Not sure if we can fix that?

**Comment by DiegoPino** [comment]
> My only question is related to the "Share this book" is it expected that the "Embed a mini book reader" says "Embed code not currently supported."?

@whikloj seems we don't have embed on the old version. I could try to remove that whole section (have to figure out how) and if i do i can also figure out how to make the finish button click event == to the one of the tiny X on the top.

**Comment by DiegoPino** [comment]
@whikloj gosh... the whole embed form is just inline html....

https://github.com/internetarchive/bookreader/blob/dd987a18dcd2a9aac93516a7605c2075de944dc0/BookReader/BookReader.js#L4150-L4209

**Comment by whikloj** [comment]
So we can fix, that... this is how I did it. Not sure if we care though?
```
  // Overrides buildShareDiv().
  IslandoraBookReader.prototype.buildShareDiv = function(jShareDiv) {
    BookReader.prototype.buildShareDiv.call(this, jShareDiv);
    var finishedDiv = jShareDiv.find( "button.share-finished" );
    if (finishedDiv.length > 0) {
      finishedDiv[0].setAttribute('onclick', "parent.jQuery.colorbox.close();");
    }
  }
```

**Comment by DiegoPino** [comment]
@whikloj jinx.. was just writing you about that.. that option seems the most clean one but won't survive a change on the HTML structure if someone decides to update in the future
Option 2 is to override completely
https://github.com/internetarchive/bookreader/blob/2.0.2/BookReader/BookReader.js#L5780-L5851

let me know what is best.

**Comment by whikloj** [comment]
I don't really care. I like leaving the Share button available but if you want to remove it we can do that too. 
Looks like the same split in code exists in version 4.2.2 so maybe a PR to them is in order.

**Comment by whikloj** [comment]
I should specify that I don't like overriding the entire form. Either fix or remove in my mind.

**Comment by DiegoPino** [comment]
@whikloj this is as close as i can get. I only removed the embed fieldset. Have to depart to day job duties, if there are other issues i would hope @patdunlavey could chime in, if not, i can eventually look at those tonight...

**Review by whikloj — APPROVED** [review]
I am good here 👍

**Comment by DiegoPino** [comment]
@whikloj also, those folks don't like external pulls too much. Depends on the day and the justification. But that $.fn is in very piece of their code, we are the ones that need to override it because of Drupal and friends, so technically (like being super permissive and tolerant) their code is not wrong (i guess)

**Comment by DigitLib** [comment]
@DiegoPino @whikloj Tried and everything WORK!!!! Even pass word from Simple Search into bookreader!  Great JOB!!! TNX!!!!

**Comment by whikloj** [comment]
@DiegoPino true.

We'll just have to be aware when we update to a later version to check that these "tweaks" are still needed and (if they are) that they haven't stopped working.

**Comment by whikloj** [comment]
Starting official 24 hour 🕐 , plus waiting for Travis.

**Comment by DiegoPino** [comment]
YU HU! Thanks @whikloj i'm sure @DigitLib and others will be happy!

**Comment by DigitLib** [comment]
@DiegoPino YES!!!! Thank YOU!!! @whikloj @giancarlobi @patdunlavey 
Now test again in ova and documenting!!!!

**Pull request merged**
